Added Eslint no-undef rule and fixed linter errors

dnguneratne opened the pull request:
Closes #1426 

This PR contains the following changes:
- Added Eslint no-undef rule to `.eslintrc.json`.
- Added `--config .eslintrc.json` flag to `package.json` `lint:js` script.
- Fixed 'unnecessary semicolon' errors highlighted by Esilnt 'no-extra-semi`.
- Fixed almost all errors due to the `no-undef` rule except for 2 in `openlibrary\plugins\openlibrary\js\ol.js`. They are:

1. `items` not defined in function `getUrlParameter` on line 4.
```js
var Browser = {
    getUrlParameter: function(key) {
        var query = window.location.search.substring(1);
        var params = query.split("&");
        if (key) {
            for (var i=0;i<params.length;i++) {
                var item = params[i].split("=");
                var val = item[1];
                if(item[0] == key){return(decodeURIComponent(val));}
            }
            return(undefined);
        }
        return(items); // <- This line
    }
```
I cannot find another reference to `items` so I'm not sure where it originates from. Was it meant to be variable like `var items = {};` like in function `getJsonFromUrl` on line 19:
```js
getJsonFromUrl: function () {
    var query = location.search.substr(1);
    var result = {}; // <- This line
    query.split("&").forEach(function(part) {
        var item = part.split("=");
        result[item[0]] = decodeURIComponent(item[1]);
    });
    return result;
},
```

2. `set_hash` is not defined on line 154.
```js
if (window.set_hash) {
    var _p = (index == 1) ? null : index;
    set_hash({"page": _p});
}
```
Again, no reference to `set_hash`.

Please let me know how I can fix these and I'll update the PR.

mekarpeles commented:
@dnguneratne thanks for your thorough work! Looking into the question you asked re: `items`

jdlrobson commented:
I think we can safely say items should be removed. I'm not sure what the original intention here was, but it throws a ReferenceError when run NOW:

```
Browser.getUrlParameter()
all.js?v=643bdbee196ef873b7175bfdde70c6ae:199 Uncaught ReferenceError: items is not defined
    at Object.getUrlParameter (all.js?v=643bdbee196ef873b7175bfdde70c6ae:199)
    at <anonymous>:1:9
getUrlParameter @ all.js?v=643bdbee196ef873b7175bfdde70c6ae:199
(anonymous) @ VM1933:1
```

Code archeology suggests this was added in f92f08bc. This code through a reference error then too. The commit message is a bit vague for that commit, but I don't see any usages of getUrlParameter when it was introduced which is usually a big red flag.


I'm also seeing no references to `getUrlParameter` in the entire codebase, so my proposal would be remove the entire function.

jdlrobson commented:
As for set_hash I'm seeing references in openlibrary/templates/lib/covers.html and openlibrary/templates/publishers/view.html and openlibrary/templates/subjects.html so please document that function to make it clear where that's coming from!

dnguneratne commented:
@jdlrobson thank you for your review. I updated the PR by removing the `getUrlParameter` function and added references to where `set_hash` is defined.

mekarpeles commented on openlibrary/plugins/openlibrary/js/ol.js:
```diff
@@ -1,19 +1,6 @@
 var startTime = new Date(); // This is used by ol.analytics.js
 
 var Browser = {
-    getUrlParameter: function(key) {
```
this looks like the right solution, thank you

mekarpeles commented:
This is looking good to me, I think we just need a rebase to resolve remaining merge conflicts

jdlrobson commented:
I can attempt to resolve these merge conflicts if @dnguneratne is unable to. Really keen to see this in as it gives us a much greater understanding of where our JS is defined!

jdlrobson commented on openlibrary/plugins/openlibrary/js/utils.js:
```diff
@@ -64,6 +64,7 @@ $.log = function() {
 function closePopup() {
     parent.jQuery.fn.colorbox.close();
 }
+=======
```
oops
